File: go/gtsdb_test.go
package gtsdb

import (
	"encoding/binary"
	"math"
	"testing"
)

// encodeBinaryFrame builds a binary frame from a map of key→points.
func encodeBinaryFrame(data map[string][]DataPoint) []byte {
	total := 4 // key count
	for k, pts := range data {
		total += 2 + len(k) + 4 + len(pts)*16
	}
	buf := make([]byte, 4, 4+total)
	binary.BigEndian.PutUint32(buf[:4], uint32(total))

	off := 4
	tmp := make([]byte, 8)
	// key count
	binary.BigEndian.PutUint32(buf[off:off+4], uint32(len(data)))
	off += 4
	buf = buf[:off]
	for k, pts := range data {
		need := 2 + len(k) + 4 + len(pts)*16
		for len(buf) < off+need {
			buf = append(buf, 0)
		}
		binary.BigEndian.PutUint16(buf[off:], uint16(len(k)))
		off += 2
		copy(buf[off:], k)
		off += len(k)
		binary.BigEndian.PutUint32(buf[off:], uint32(len(pts)))
		off += 4
		for _, dp := range pts {
			binary.BigEndian.PutUint64(tmp, uint64(dp.Timestamp))
			copy(buf[off:], tmp)
			off += 8
			binary.BigEndian.PutUint64(tmp, math.Float64bits(dp.Value))
			copy(buf[off:], tmp)
			off += 8
		}
	}
	return buf
}

func TestParseBinaryMulti(t *testing.T) {
	input := map[string][]DataPoint{
		"sensor1": {
			{Key: "sensor1", Timestamp: 1700000000, Value: 1.5},
			{Key: "sensor1", Timestamp: 1700000001, Value: 3.0},
		},
		"sensor2": {
			{Key: "sensor2", Timestamp: 1700000100, Value: -42.75},
		},
	}
	frame := encodeBinaryFrame(input)
	result, err := parseBinaryMulti(frame[4:]) // skip length prefix
	if err != nil {
		t.Fatal(err)
	}
	if len(result) != 2 {
		t.Fatalf("expected 2 keys, got %d", len(result))
	}
	pts1 := result["sensor1"]
	if len(pts1) != 2 {
		t.Fatalf("sensor1: expected 2 points, got %d", len(pts1))
	}
	if pts1[0].Timestamp != 1700000000 || pts1[0].Value != 1.5 {
		t.Errorf("sensor1[0]: {%d, %f}", pts1[0].Timestamp, pts1[0].Value)
	}
	if pts1[1].Timestamp != 1700000001 || pts1[1].Value != 3.0 {
		t.Errorf("sensor1[1]: {%d, %f}", pts1[1].Timestamp, pts1[1].Value)
	}
	pts2 := result["sensor2"]
	if len(pts2) != 1 {
		t.Fatalf("sensor2: expected 1 point, got %d", len(pts2))
	}
	if pts2[0].Timestamp != 1700000100 || pts2[0].Value != -42.75 {
		t.Errorf("sensor2[0]: {%d, %f}", pts2[0].Timestamp, pts2[0].Value)
	}
}

func TestParseBinaryMultiEmpty(t *testing.T) {
	frame := encodeBinaryFrame(map[string][]DataPoint{})
	result, err := parseBinaryMulti(frame[4:])
	if err != nil {
		t.Fatal(err)
	}
	if len(result) != 0 {
		t.Errorf("expected empty, got %d keys", len(result))
	}

	// nil/empty input
	r, err := parseBinaryMulti(nil)
	if err != nil || r != nil {
		t.Error("nil should return nil")
	}
	r, err = parseBinaryMulti([]byte{})
	if err != nil || r != nil {
		t.Error("empty should return nil")
	}
}

func TestParseBinaryMultiSpecialValues(t *testing.T) {
	input := map[string][]DataPoint{
		"k": {
			{Key: "k", Timestamp: 0, Value: 0},
			{Key: "k", Timestamp: -1, Value: math.Inf(1)},
			{Key: "k", Timestamp: 1<<62, Value: math.NaN()},
		},
	}
	frame := encodeBinaryFrame(input)
	result, err := parseBinaryMulti(frame[4:])
	if err != nil {
		t.Fatal(err)
	}
	pts := result["k"]
	if len(pts) != 3 {
		t.Fatalf("expected 3 points, got %d", len(pts))
	}
	if pts[0].Timestamp != 0 || pts[0].Value != 0 {
		t.Errorf("zero: {%d, %f}", pts[0].Timestamp, pts[0].Value)
	}
	if pts[1].Timestamp != -1 || !math.IsInf(pts[1].Value, 1) {
		t.Errorf("neg/inf: {%d, %f}", pts[1].Timestamp, pts[1].Value)
	}
	if pts[2].Timestamp != 1<<62 || !math.IsNaN(pts[2].Value) {
		t.Errorf("large/nan: {%d, %f}", pts[2].Timestamp, pts[2].Value)
	}
}

func TestParseBinaryKey(t *testing.T) {
	input := map[string][]DataPoint{
		"only": {{Key: "only", Timestamp: 42, Value: 99.9}},
	}
	frame := encodeBinaryFrame(input)
	pts, err := parseBinaryKey(frame[4:])
	if err != nil {
		t.Fatal(err)
	}
	if len(pts) != 1 {
		t.Fatalf("expected 1 point, got %d", len(pts))
	}
	if pts[0].Timestamp != 42 || pts[0].Value != 99.9 {
		t.Errorf("got {%d, %f}", pts[0].Timestamp, pts[0].Value)
	}
}

func TestParseBinaryKeyEmpty(t *testing.T) {
	frame := encodeBinaryFrame(map[string][]DataPoint{})
	pts, err := parseBinaryKey(frame[4:])
	if err != nil {
		t.Fatal(err)
	}
	if pts != nil {
		t.Errorf("expected nil, got %v", pts)
	}
}

func TestReadBinaryFrame(t *testing.T) {
	// This tests the frame splitting logic indirectly via parse
	// The actual io.ReadFull is tested in integration
}

func TestDataPointStruct(t *testing.T) {
	dp := DataPoint{Key: "test", Timestamp: 100, Value: 3.14}
	if dp.Key != "test" || dp.Timestamp != 100 || dp.Value != 3.14 {
		t.Error("struct fields mismatch")
	}
}

File: js/gtsdb.test.js
const { GTSDBClient } = require('./gtsdb.js');
const assert = require('assert');

// Helper: build a binary frame identical to Go's binary.go
function encodeBinaryFrame(data) {
  const parts = [];
  const keys = Object.keys(data);

  // Key count
  const keyCountBuf = Buffer.alloc(4);
  keyCountBuf.writeUInt32BE(keys.length, 0);
  parts.push(keyCountBuf);

  for (const key of keys) {
    const pts = data[key];
    // Key length + key
    const keyLenBuf = Buffer.alloc(2);
    keyLenBuf.writeUInt16BE(Buffer.byteLength(key), 0);
    parts.push(keyLenBuf);
    parts.push(Buffer.from(key, 'utf8'));

    // Point count
    const ptCountBuf = Buffer.alloc(4);
    ptCountBuf.writeUInt32BE(pts.length, 0);
    parts.push(ptCountBuf);

    for (const pt of pts) {
      const tsBuf = Buffer.alloc(8);
      tsBuf.writeBigInt64BE(BigInt(pt.timestamp), 0);
      parts.push(tsBuf);

      const valBuf = Buffer.alloc(8);
      valBuf.writeDoubleBE(pt.value, 0);
      parts.push(valBuf);
    }
  }

  const dataBuf = Buffer.concat(parts);
  const lenBuf = Buffer.alloc(4);
  lenBuf.writeUInt32BE(dataBuf.length, 0);
  return Buffer.concat([lenBuf, dataBuf]);
}

describe('GTSDBClient', () => {
  describe('_parseBinaryMulti', () => {
    const client = new GTSDBClient('localhost', 5555);

    it('parses multiple keys with multiple points', () => {
      const input = {
        sensor1: [
          { key: 'sensor1', timestamp: 1700000000, value: 1.5 },
          { key: 'sensor1', timestamp: 1700000001, value: 3.0 },
        ],
        sensor2: [
          { key: 'sensor2', timestamp: 1700000100, value: -42.75 },
        ],
      };
      const frame = encodeBinaryFrame(input);
      const result = client._parseBinaryMulti(frame.slice(4)); // skip len prefix
      assert.strictEqual(Object.keys(result).length, 2);
      assert.strictEqual(result.sensor1.length, 2);
      assert.strictEqual(result.sensor1[0].timestamp, 1700000000);
      assert.strictEqual(result.sensor1[0].value, 1.5);
      assert.strictEqual(result.sensor1[1].timestamp, 1700000001);
      assert.strictEqual(result.sensor1[1].value, 3.0);
      assert.strictEqual(result.sensor2.length, 1);
      assert.strictEqual(result.sensor2[0].timestamp, 1700000100);
      assert.strictEqual(result.sensor2[0].value, -42.75);
    });

    it('parses empty data', () => {
      const frame = encodeBinaryFrame({});
      const result = client._parseBinaryMulti(frame.slice(4));
      assert.strictEqual(Object.keys(result).length, 0);
    });

    it('handles special float values', () => {
      const input = {
        k: [
          { key: 'k', timestamp: 0, value: 0 },
          { key: 'k', timestamp: -1, value: Infinity },
          { key: 'k', timestamp: Number.MAX_SAFE_INTEGER, value: NaN },
        ],
      };
      const frame = encodeBinaryFrame(input);
      const result = client._parseBinaryMulti(frame.slice(4));
      const pts = result.k;
      assert.strictEqual(pts.length, 3);
      assert.strictEqual(pts[0].timestamp, 0);
      assert.strictEqual(pts[0].value, 0);
      assert.strictEqual(pts[1].timestamp, -1);
      assert(pts[1].value === Infinity);
      assert(Number.isNaN(pts[2].value));
    });

    it('returns empty object for short buffers', () => {
      assert.deepStrictEqual(client._parseBinaryMulti(Buffer.alloc(0)), {});
    });
  });

  describe('constructor', () => {
    it('creates client with default state', () => {
      const c = new GTSDBClient('localhost', 5555);
      assert.strictEqual(c.host, 'localhost');
      assert.strictEqual(c.port, 5555);
      assert.strictEqual(c.connected, false);
      assert.strictEqual(c.buffer, '');
    });
  });
});
